[KEP-1282]: Update evictionState of WorkloadStatus (#5238)
kubernetes-sigs/kueue: keps/1282-pods-ready-requeue-strategy/README.md

}
```

> Add a new field `evictionState` to the WorkloadStatus struct to track the number of evictions by  reason:

Reviewer: ```suggestion
Add a new field, `evictionState`, to the WorkloadStatus struct to track the number of evictions by reason:
```

Add a new field `evictionState` to the WorkloadStatus struct to track the number of evictions by  reason:

```go
type WorkloadStatus struct {
	// evictionStats tracks eviction statistics by type and reason.
    // evictionState tracks eviction statistics by reason.
    //
    // +optional
    // +listType=map
    // +listMapKey=type
    // +patchStrategy=merge
    // +patchMergeKey=type
    EvictionStats []Eviction `json:"evictionStats,omitempty"`
    EvictionStates []WorkloadEvictionState `json:"evictionStates,omitempty"`
}

type Eviction struct {
    // type identifies the eviction type.
    //
    // +required
    // +kubebuilder:validation:Required
    // +kubebuilder:validation:Pattern=`^([a-z0-9]([-a-z0-9]*[a-z0-9])?(\.[a-z0-9]([-a-z0-9]*[a-z0-9])?)*/)?(([A-Za-z0-9][-A-Za-z0-9_.]*)?[A-Za-z0-9])$`
    // +kubebuilder:validation:MaxLength=316

>     Type string `json:"type"`

Reviewer 2: Actually, having the "type" and "reason" at the metric level also allows us to easily aggregate the metrics in grafana at both levels.

>     Type string `json:"type"`

Reviewer 2: Actually, seeing that the proposal migrates to be more generic, I think instead of introducing a new metric like https://github.com/kubernetes-sigs/kueue/pull/5127 `kueue_wait_for_pods_ready_workload_eviction_total` I would suggest to introduce the following metric, adding the "detailedReason".

So:
- EvictedWorkloadsTotal -> fix to count workloads only once.
- EvictionsTotal -> would count eviction events as such

Both metrics would have the "detailedReason". In the API:
- "type" -> "reason"
- "reason" -> "detailedReason"

```
	EvictedWorkloadsTotal = prometheus.NewCounterVec(
		prometheus.CounterOpts{
			Subsystem: constants.KueueName,
			Name:      "evicted_workloads_total",
			Help: `The number of evicted workloads per 'cluster_queue',
The label 'reason' can have the following values:
- "Preempted" means that the workload was evicted in order to free resources for a workload with a higher priority or reclamation of nominal quota.
- "PodsReadyTimeout" means that the eviction took place due to a PodsReady timeout.
- "AdmissionCheck" means that the workload was evicted because at least one admission check transitioned to False.
- "ClusterQueueStopped" means that the workload was evicted because the ClusterQueue is stopped.
- "Deactivated" means that the workload was evicted because spec.active is set to false`,
		}, []string{"cluster_queue", "reason", "detailed_reason"},
	)
```
```
	EvictionsTotal = prometheus.NewCounterVec(
		prometheus.CounterOpts{
			Subsystem: constants.KueueName,
			Name:      "evictions_total",
			Help: `The number of evictions per 'cluster_queue',
The label 'reason' can have the following values:
- "Preempted" means that the workload was evicted in order to free resources for a workload with a higher priority or reclamation of nominal quota.
- "PodsReadyTimeout" means that the eviction took place due to a PodsReady timeout.
- "AdmissionCheck" means that the workload was evicted because at least one admission check transitioned to False.
- "ClusterQueueStopped" means that the workload was evicted because the ClusterQueue is stopped.
- "Deactivated" means that the workload was evicted because spec.active is set to false`,
		}, []string{"cluster_queue", "reason", "detailed_reason"},
	)
```

>     Type string `json:"type"`

Reviewer 3: I like the idea of generic metric. And if the current metrics have both `reason` and `detailedReason`, I would like to stick with it for the sake of consistency

>     Type string `json:"type"`

Author: I am missing some details here...
1. Should both metrics count evictions per `cluster_queue` ?
2. Should the `evictions_total` is `cluster_queue` independent?
If not, why we need both?
3. If both are per `cluster_queue` then `EvictedWorkloadsTotal -> fix to count workloads only once.` should rather be a gauge to count up and down and `EvictionsTotal -> would count eviction events as such` - counts only up?

>     Type string `json:"type"`

Reviewer 4: > Actually, having the "type" and "reason" at the metric level also allows us to easily aggregate the metrics in grafana at both levels of detail.

At prometheus metrics level, I'm ok with including type. However, at the Workload API level, I have some objections.
The state machine should be defined in a single place, and we should expose only additional information via another state field like `evictionState`. Otherwise, the state machine will be more complex across multiple fields and bring us uncontrollable state handling.

>     Type string `json:"type"`

Reviewer 2: I think the "type" could have been misleading. The metric we have already has "reason" as one of the following:
- "Preempted" means that the workload was evicted in order to free resources for a workload with a higher priority or reclamation of nominal quota.
- "PodsReadyTimeout" means that the eviction took place due to a PodsReady timeout.
- "AdmissionCheck" means that the workload was evicted because at least one admission check transitioned to False.
- "ClusterQueueStopped" means that the workload was evicted because the ClusterQueue is stopped.
- "Deactivated" means that the workload was evicted because spec.active is set to false`,

I think in this feature we need something like "detailedReason" to break down "PodsReadyTimeout" into two subcategories.

>     Type string `json:"type"`

Author: OK, then the actual difference is one counts workload evictions per cq and the other counts all evictions per cq?
What else do we evict apart from workloads?

>     Type string `json:"type"`

Reviewer 4: @mimowo Is this feature objective adding human-readable information or machine-readable information or both?

>     Type string `json:"type"`

Reviewer 2: machine readable

>     Type string `json:"type"`

Reviewer 2: > In that case, shoulnd't we have 2 different reasons instead of a single PodsReadyTimeout?

For this feature (metric) we want to distinguish StartupTimeout and RecoveryTimeout. I think since the metric is going to be generic then for preemption we would need to distinguish the following: https://github.com/kubernetes-sigs/kueue/blob/4cafd5fbad492983c9e3b8fde2dea2823712d44f/apis/kueue/v1beta1/workload_types.go#L517-L534

Would you like to mesh all the reasons into one big enum, making sure the reasons from different features don't conflict?

>     Type string `json:"type"`

Reviewer 4: Ah, I imagined that you want to use `PodsReadyTimeout` in the `evictionStates.reason` for WorkloasStatus resource. 
If we introduce the `detailed_reason` only to Prometheus metrics, I'm fine with that.

>     Type string `json:"type"`

Reviewer 2: Well, indeed I imagined:
```yaml
- reason: PodsReadyTimeout
  detailedReason: StartupTimeout
  count: 4
- reason: Preempted
  detailedReason: InCohortReclamation
  count: 2
- reason: AdmissionCheck
  detailedReason: // not available yet
```
Then, to put the information into the metrics. It would allow to easily have a metric as the current `EvictedWorkloadsTotal` but just extended by "detailedReason".

>     Type string `json:"type"`

Reviewer 4: Let me put the reason why I am keeping urgue not to introduduce those fields to Workload status: https://docs.google.com/presentation/d/1_4PCAne1b1NtggdeemBkFmJvVctr4b_Z5n1rsJsmg-Q/edit?usp=sharing

>     Type string `json:"type"`

Reviewer 2: Yes. there is some tranformation needed, but not to the "detailedReason" part, which would be 1:1 taken from the PodsReadyTimeout condition. I think some transformation is done for the preexisting reason in the preexisting metric `evicted_workloads_total`. 

Yuki, what is your proposal?

>     Type string `json:"type"`

Reviewer 4: Before something, could you tell me about the missing capabilities, as I mentioned on page 4?

>     Type string `json:"type"`

Reviewer 4: > Yes. there is some tranformation needed, but not to the "detailedReason" part, which would be 1:1 taken from the PodsReadyTimeout condition.

The problem with multi-time transformations is that they need to be considered to keep consistency between conditions and evictionStetes which considering the behind criteria for each condition.

>     Type string `json:"type"`

Reviewer 3: Could you expand on how do you understand "two times stretched state" and why is this a problem? The `evictionState` will only changed by increasing counter, adding a new reason when Workloads gets evicted? The interaction is only one way, and I think it's fairly simply

I see two main capabilities of this feature:
1. Have a metrics that counts how many different workloads have been evicted (in particular, by waitForPodsReady)
2. From the user perspective, be able to easily find workloads that have been evicted in the past (in particular by waitForPodsReady)

I believe we need status API for that, as keeping the state in memory won't work on restart, and it doesn't meet the 2nd requirement

>     Type string `json:"type"`

Reviewer 5: I think that the counter field for eviction/attempts will have couple other applications. For example we can store information which RF we tried (and failed) in the past and possibly try a different one the next time. So, most likely, PR is not only for the metric, although this is the most immediate application.

>     Type string `json:"type"`

Reviewer 4: > I think that the counter field for eviction/attempts will have couple other applications. For example we can store information which RF we tried (and failed) in the past and possibly try a different one the next time. So, most likely, PR is not only for the metric, although this is the most immediate application.

I can understand that this feature is useful. The most impactful problem is the `evictionState` has multiple steps as one state like `WaitForStart` -> `PodsReadyTimeout` -> `Eviction`

>     Type string `json:"type"`

Reviewer 2: >  The most impactful problem is the evictionState has multiple steps as one state like WaitForStart -> PodsReadyTimeout -> Eviction

Could you elaborate on why this is a problem? The implementation PR  is relatively simple.

I would like to see some alternative proposal to satisfy the feature needs. 

I'm not a big fun of this, but in order to make progress we could have:
```yaml
evictionState:
- reason: WaitForStart
  count: 
```
So in API we would have just one API field, and we would implicitly know that WaitForStart means PodsReadyTimeout when setting the "reason" and "detailed_reason" labels.

>     Type string `json:"type"`

Reviewer 4: The discussion results are stored in https://github.com/kubernetes-sigs/kueue/pull/5238#discussion_r2093201954
Please see it if anyone is interested in this discussion.

>     // count tracks the number of evictions for the reason.

Reviewer: ```suggestion
    // count tracks the number of evictions for this reason.
```

}
```

Add a new field, `evictionState`, to the WorkloadStatus struct to track the number of evictions by reason:

```go
type WorkloadStatus struct {
    // evictionState tracks eviction statistics by reason.
    //
    // +optional
    // +listType=map
    // +listMapKey=type
    // +patchStrategy=merge
    // +patchMergeKey=type
    EvictionStates []WorkloadEvictionState `json:"evictionStates,omitempty"`
}

type WorkloadEvictionState struct {
    // reason specifies the programmatic identifier for the eviction cause.
    //
    // +required
    // +kubebuilder:validation:Required
    // +kubebuilder:validation:Enum=WaitForStart;WaitForRecovery

>     Reason string `json:"reason"`

Reviewer 4: ```suggestion
    Reason WorkloadEvictionStateReason `json:"reason"`
```

@mszadkow could we use the dedicated type to prevent assigning arbitrary reason code base? Do you see any problem to introduce dedicated typed reason?

Author: Could be done, the only counter argument I have, is that we will duplicate existing Workload reasons.

Author: https://github.com/kubernetes-sigs/kueue/pull/5211#discussion_r2086246430
That speaks to me too.
Let's make it separate enum

}
```

Add a new field, `evictionState`, to the WorkloadStatus struct to track the number of evictions by reason:

```go
type WorkloadStatus struct {
    // evictionState tracks eviction statistics by reason.
    //
    // +optional
    // +listType=map
    // +listMapKey=reason
    // +patchStrategy=merge
    // +patchMergeKey=reason
    EvictionStates []WorkloadEvictionState `json:"evictionStates,omitempty"`
}

> type WorkloadEvictionState struct {

Reviewer 5: Eviction limits the applciation to eviction only. Can we make it a little bit more generic?

Reviewer 2: We can try, but I'm worried we can over-engineer this. Do you have any specific proposal?

I think given the data is going to supply the EvictedWorkloadOnce metric it makes sense to keep the name scoped to "eviction".

We can always revisit the API for v1beta2.

Reviewer 5: What I'm suggesting is just a minor rename. Instead of calling it eviction state maybe we can call it, for example, scheduling state - almost the same, but leaves other applications open.

Reviewer 2: but it embeds the list of eviction reasons, directly, since this is `EvictionStates []WorkloadEvictionState`. 

Maybe we could add another layer of nesting to accommodate for it
```yaml
schedulingState:
  // with potential other info
  evictionState:
  - reason: ...
     detailedReason
     count: 
```

Reviewer 5: OK, i understand calling the field evictionState and making it an array. What I'm trying to point out is that the structure has no field very specific to eviction it is reason, details and count. So I was thinking about removing eviction from the struct name to allow future reuse.

Reviewer 2: I synced with @mwielgus and we concluded on the following option as preferred:

```yaml
schedulingState:
  // potential place to include admissionStats, etc.
  evictionStats:  
  - reason: PodsReadyTimeout  // From Evicted.reason
    detailedReason: WaitForStart  // From PodsReady.reason
    count: 3
```
but skipping `schedulingState` is also acceptable.

    // reason specifies the programmatic identifier for the eviction cause.
    //
    // +required
    // +kubebuilder:validation:Required
    // +kubebuilder:validation:MaxLength=316
    Reason string `json:"reason"`

    // detailedReason specifies a finer-grained explanation that complements the eviction cause.
    // This may be an empty string.
    //
    // +required
    // +kubebuilder:validation:Required
    // +kubebuilder:validation:MaxLength=316
    DetailedReason string `json:"detailedReason"`

    // count tracks the number of evictions for this reason and detailed reason.

>     //

Reviewer 5: I guess we will need LastUpdate soon. Anyway, this can wait until we have a PR that uses it.

}
```

Add a new field, `evictionState`, to the WorkloadStatus struct to track the number of evictions by reason:

```go
type WorkloadStatus struct {
    // evictionState tracks eviction statistics by reason.
    //
    // +optional
    // +listType=map
    // +listMapKey=type
    // +patchStrategy=merge
    // +patchMergeKey=type
    EvictionStates []WorkloadEvictionState `json:"evictionStates,omitempty"`
}

type WorkloadEvictionStateReason string

const (
	Preempted WorkloadEvictionStateReason = "Preempted"
	PodsReadyTimeout WorkloadEvictionStateReason = "PodsReadyTimeout"
	AdmissionCheck WorkloadEvictionStateReason = "AdmissionCheck"
	ClusterQueueStopped WorkloadEvictionStateReason = "ClusterQueueStopped"
	Deactivated WorkloadEvictionStateReason = "Deactivated"
)

type WorkloadEvictionState struct {
    // reason specifies the programmatic identifier for the eviction cause.
    //
    // +required
    // +kubebuilder:validation:Required
    // +kubebuilder:validation:Enum={Preempted,PodsReadyTimeout,AdmissionCheck,ClusterQueueStopped,Deactivated}
    Reason WorkloadEvictionStateReason `json:"reason"`
    

>     // count tracks the number of evictions for this reason.

Reviewer 2: Don't we need also the "DetailedReason" field?

Author: Yes, done.

}
```

Add a new field, `schedulingState`, to the WorkloadStatus struct to track the number of scheduling statistics.
SchedulingState serves as a history of the Workload thus will not be reset.

> First sub-resource to be a part of the `schedulingState` will be `evictionStates` a struct to track the number of evictions by reason.

Reviewer 4: ```suggestion
First sub field to be a part of the `schedulingState` will be `evictionStates` a struct to track the number of evictions by reason.
```

The `sub-resource` has a dedicated technical term which indicates `status` and `scale`, something else.

```go
type WorkloadStatus struct {
	// schedulingState tracks scheduling statistics (e.g. evictionStates)
	//
	// +optional
	SchedulingState SchedulingState `json:"schedulingState,omitempty"`
}

type SchedulingState struct {
	// evictionState tracks eviction statistics by reason.
	//
	// +optional
	// +listType=map
	// +listMapKey=reason
	// +patchStrategy=merge

> 	// +patchMergeKey=reason

Reviewer 4: Does this `+ patchMergeKey ` and `+listMapKey` work well? I am suspecting we have conflicts in the following:
Did you check it?

```yaml
schedulingState:
  evictionStats:  
  - reason: PodsReadyTimeout  // From Evicted.reason
    underlyingCause: WaitForStart // From PodsReady.reason
    count: 3
  - reason: PodsReadyTimeout  // From Evicted.reason
    underlyingCause: WaitForRecovery // From PodsReady.reason
    count: 3
```

Author: Yes, you are correct I will change it to composite key

Author: We have basically 2 options here.
Either we introduce an extra field.g. `uniqueKey` as combination of `reason` and `underlyingCause` or we assume `underlyingCause` is more unique than `reason`.
wdyt? @mimowo @tenzen-y

Reviewer 4: IIUC, we can specify the multiple map keys in the following. Did you see any generation errors?

https://kubernetes.io/docs/reference/using-api/server-side-apply/#merge-strategy

```go
	// +optional
	// +listType=map
	// +listMapKey=reason
	// +listMapkey=underlyingCause
	// +patchStrategy=merge
	// +patchMergeKey=reason
	// +patchMergeKey=underlyingCause
```

Author: Nope, no errors

}
```

Add a new field, `evictionState`, to the WorkloadStatus struct to track the number of evictions by reason:

```go
type WorkloadStatus struct {
    // evictionState tracks eviction statistics by reason.
    //
    // +optional
    // +listType=map
    // +listMapKey=type
    // +patchStrategy=merge

>     // +patchMergeKey=type

Author: // +optional
	// +listType=map
	// +listMapKey=reason
	// +patchStrategy=merge
	// +patchMergeKey=reason

    EvictionStates []WorkloadEvictionState `json:"evictionStates,omitempty"`
}

type WorkloadEvictionStateReason string

const (
	Preempted WorkloadEvictionStateReason = "Preempted"
	PodsReadyTimeout WorkloadEvictionStateReason = "PodsReadyTimeout"
	AdmissionCheck WorkloadEvictionStateReason = "AdmissionCheck"
	ClusterQueueStopped WorkloadEvictionStateReason = "ClusterQueueStopped"
	Deactivated WorkloadEvictionStateReason = "Deactivated"
)

type WorkloadEvictionState struct {
    // reason specifies the programmatic identifier for the eviction cause.
    //
    // +required
    // +kubebuilder:validation:Required
    // +kubebuilder:validation:Enum={Preempted,PodsReadyTimeout,AdmissionCheck,ClusterQueueStopped,Deactivated}

>     Reason WorkloadEvictionStateReason `json:"reason"`

Reviewer 2: That would be my feeling too. I would be ok to start with string and refactor to use enum. cc @tenzen-y @PBundyra

>     Reason WorkloadEvictionStateReason `json:"reason"`

Reviewer 3: >  I would be ok to start with string and refactor to use enum

I'm fine with this approach

}
```

> Add a new field, `schedulingState`, to the WorkloadStatus struct to track the number of scheduling statistics.

Reviewer 2: Since the API is going to be more generic than the metric (and expanded in the future), let me consider alternative naming:
1. schedulingState, evictionStates (as currently)
2. schedulingStats, evictionStats
3. schedulingEvents, evictionEvents
4. schedulingHistory, evictionHistory
5. schedulingLog, evictionLog

I'm not a big fan of:
_  `schedulingState`, because "state" suggests to me this is "current" moment, but the structure is meant for historical data. 
- `schedulingStats` - because id does not capture the intention - these are not statistics like avg, std dev, these are just data to compute some stats on. 
- `schedulingEvents` - because events has already a formal meaning in Kubernetes, so could be constraining if we want to use it beyond events.

So, my preference is either `schedulingLog` or `schedulingHistory`. I'm slightly leaning towards `schedulingLog` as a bit more "relaxed".

Let me know what you think @tenzen-y @PBundyra @mwielgus

Reviewer 3: I'm okay with both 2 and 4

Reviewer 4: I think `2` is better since `Log` and `History` indicate the raw record of past scheduling. The `Stats` means the aggregation and summarization of log or history based on some points.

Reviewer 4: In that case, `schedulingStats.eviction` looks better, IMO.
For sure, the `stats` indicates `statistics` like avg, std dev in the mathematical field.
However, the word ' stats` means the information based on something, which is not limited in dedicated signs.

For `schedulingLogSummary` means more modified information rather than `statics`. I meant `Summary` indicates analized information by someone (kueue-controller-manager) PoV.

Reviewer 2: Ok, let's go with `schedulingStats.eviction`. @mszadkow please update the KEP and implementation accordingly.

Author: Sure, I am on it

SchedulingState serves as a history of the Workload thus will not be reset.
First sub field to be a part of the `schedulingState` will be `evictionStates` a struct to track the number of evictions by reason.

```go
type WorkloadStatus struct {
	// schedulingState tracks scheduling statistics (e.g. evictionStates)
	//
	// +optional

> 	SchedulingState SchedulingState `json:"schedulingState,omitempty"`

Reviewer 4: ```suggestion
	SchedulingState *SchedulingState `json:"schedulingState,omitempty"`
```

Let's use an pointer field.

}

type SchedulingStats struct {
	// eviction tracks eviction statistics by reason and underlyingCause.
	//
	// +optional
	// +listType=map
	// +listMapKey=reason
	// +listMapKey=underlyingCause
	// +patchStrategy=merge
	// +patchMergeKey=reason
	// +patchMergeKey=underlyingCause

> 	Eviction []WorkloadSchedulingStatsEvicition `json:"eviction,omitempty"`

Reviewer 4: ```suggestion
	Evictions []WorkloadSchedulingStatsEvicition `json:"evictions,omitempty"`
```

Shouldn't we use the `evictions` since this can accommodate the multiple eviction stats?
/hold

Author: I thought about that, but I read this like "scheduling statistics related to eviction".
Then I thought keeping eviction (singular form) was ok here.
Wdyt @mimowo @PBundyra ?

Author: but I can't easily change it

Reviewer 2: I think both singlular and plural have some merit. I would be slightly leaning to Evictions as it is a list.

Reviewer 4: IIUC, Workload has eviction statistics every for reason and underlyingCause.
This is because I recommended using `evicitons`. We are going to store the statistics for eviction for each reason.
In other words, the workload has eviction statistics for PodsReadyTimeout+WaitForStart and statistics for PodsReadyTimeout+WaitForRecovery in the following:

```yaml
schedulingStats:
  evictions:  
  - reason: PodsReadyTimeout  // From Evicted.reason
    underlyingCause: WaitForStart // From PodsReady.reason
    count: 3
    lastTime: xyz // Not for Now
  - reason: PodsReadyTimeout  // From Evicted.reason
    underlyingCause: WaitForRecovery // From PodsReady.reason
    count: 3
    lastTime: xyz // Not for Now
```

Reviewer 2: ok, @mszadkow please adjust accordingly

Author: done, please check again @tenzen-y

}

> type WorkloadEvictionState struct {

Reviewer 2: ```suggestion
type WorkloadSchedulingStatsEvicition struct {
```

> type WorkloadSchedulingStatsEvicition struct {

Reviewer 2: nit WorkloadSchedulingStatsEvicitions
to align with the field

Reviewer 4: Couldn't we keep `WorkloadSchedulingStatsEvicition`? Because evicions are set of `WorkloadSchedulingStatsEvicition`.

[deprecation-policy]: https://kubernetes.io/docs/reference/using-api/deprecation-policy/
-->

### Changes to metrics

> Extend an existing `evicted_workloads_total` metric to include also `detailedReason`.

Reviewer 4: ```suggestion
Extend an existing `evicted_workloads_total` metric to include also `detailed_reason`.
```

> Introduce a new metric `evictions_total` to count total number of evictions per 'cluster_queue'.

Author: After discussion with Michal I see it like this now:
Number of evictions total with reason to an eviction per cluster queue. Counts all events.
Number of evicted workloads per cluster_queue. Count only one event per workload.

> Introduce a new metric `evictions_total` to count total number of evictions per 'cluster_queue'.

Reviewer 2: having a second thought on that. maybe it is better to keep the existing metric as is, just add a new one 'evicted_workloads_count_once_total'. WDYT @mszadkow @tenzen-y ?

> Introduce a new metric `evictions_total` to count total number of evictions per 'cluster_queue'.

Author: We want this metric to convey the idea of "once per workload-reason pair", I think it will be just enough.

> Introduce a new metric `evictions_total` to count total number of evictions per 'cluster_queue'.

Author: Sure let's have it all together here:

Workload.Status.evictionStates:
- reason: PodsReadyTimeout
  detailedReason: StartupTimeout
  count: 4
- reason: Preempted
  detailedReason: InCohortReclamation
  count: 2
- reason: AdmissionCheck
  detailedReason: // not available yet
  count: 3

Metrics:

{"kueue_evicted_workloads_total", cq, "9"},

{"kueue_evicted_workloads_once_total", cq, "PodsReadyTimeout", "StartupTimeoutReason", "1"},
{"kueue_evicted_workloads_once_total", cq, "Preempted", "InCohortReclamation", "1"},
{"kueue_evicted_workloads_once_total", cq, "AdmissionCheck", nil, "1"},

```go

> 	EvictedWorkloadsTotal = prometheus.NewCounterVec(

Reviewer 2: As mentioned in the other comment, to minimize the diff, and avoid changing the existing metric I would actually propose to keep `evicted_workloads_total`, and just add `evicted_workloads_once_total`.

Reviewer 2: So we will not have "reason" and "detailedReason", just "reason"?

Author: No, let's have `reason` and `detailed_reason`, one will be typed string the other is just a string.

Author: However I  don't have opinion on `detailed_reason` in `EvictedWorkloadsTotal`.
Maybe it's better to leave it as is

Reviewer 2: Yes, to make the changes minimal I'm ok to do not add the `detailed_reason` label to the pre-existing `EvictedWorkloadsTotal`. My goal was to make sure this is possible in the future, and the API looks consistent. 

For now, I would just keep `EvictedWorkloadsTotal` as is and introduce `EvictedWorkloadsOnceTotal`.
